driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -293,10 +320,48 @@ private PooledConnection getPooledConnectionImmediately() {
     }
 
     private MongoTimeoutException createTimeoutException(final Timeout timeout) {
-        return new MongoTimeoutException(format("Timed out after %s while waiting for a connection to server %s.",
-                                                timeout.toUserString(), serverId.getAddress()));
+        int numPinnedToCursor = pinnedStatsManager.getNumPinnedToCursor();
+        int numPinnedToTransaction = pinnedStatsManager.getNumPinnedToTransaction();
+        if (numPinnedToCursor == 0 && numPinnedToTransaction == 0) {
+            return new MongoTimeoutException(format("Timed out after %s while waiting for a connection to server %s.",
+                    timeout.toUserString(), serverId.getAddress()));
+        } else {
+            int maxSize = settings.getMaxSize();
+            int numInUse = pool.getInUseCount();
+            /* At this point in an execution we consider at least one of `numPinnedToCursor`, `numPinnedToTransaction` to be positive.
+             * `numPinnedToCursor`, `numPinnedToTransaction` and `numInUse` are not a snapshot view,
+             * but we still must maintain the following invariants:
+             * - numInUse > 0
+             *     we consider at least one of `numPinnedToCursor`, `numPinnedToTransaction` to be positive,
+             *     so if we observe `numInUse` to be 0, we have to estimate it based on `numPinnedToCursor` and `numPinnedToTransaction`;
+             * - numInUse < maxSize
+             *     `numInUse` must not exceed the limit in situations when we estimate `numInUse`;
+             * - numPinnedToCursor + numPinnedToTransaction <= numInUse
+             *     otherwise the numbers do not make sense.
+             */
+            if (numInUse == 0) {
+                numInUse = Math.min(
+                        numPinnedToCursor + numPinnedToTransaction, // must be at least a big as this sum but not bigger than `maxSize`
+                        maxSize);
+            }
+            numPinnedToCursor = Math.min(
+                    numPinnedToCursor, // prefer the observed value, but it must not be bigger than `numInUse`
+                    numInUse);
+            numPinnedToTransaction = Math.min(
+                    numPinnedToTransaction, // prefer the observed value, but it must not be bigger than `numInUse` - `numPinnedToCursor`
+                    numInUse - numPinnedToCursor);
+            int numOtherInUse = numInUse - numPinnedToCursor - numPinnedToTransaction;
+            assertTrue(numOtherInUse >= 0);
+            assertTrue(numPinnedToCursor + numPinnedToTransaction + numOtherInUse <= maxSize);
+            return new MongoTimeoutException(format("Timed out after %s while waiting for a connection to server %s. Details: "
```

Comment:
I don't get why this is so complex - do users *really* need to know all this information or are we just using it as a way to test?

> Timed out after 10000 ms while waiting for a connection to server localhost. Details:  maxPoolSize: 30, connections in use by cursors: 20, connections in use by transactions: 5, connections in use by other operations: 5"

Is that any more useful than: 

> Timed out after 10000 ms while waiting for a connection to server localhost. All connections in use with max pool size: 30.

Reply:
> All connections in use with max pool size: 30.

It is worth pointing out that a checkout request may also timeout because of `maxConnecting`, not only because of `maxSize`.

P.S. I am also not sure that complexity imposed by the spec is worth the intent.

Reply:
Resolving... As this behavior is required by the specification, we would have to change the specification first before removing this code.